### PR TITLE
Logging Reform. Integrate npmlog, char-spinner. Resolves gh-607

### PR DESCRIPTION
**Pull request by rwaldron**
This replaces our logging with npmlog and introduces a standardized integration of "char-spinner" which is used to indicate "something is happening".

This reformed logging system will allow making all logging levels configurable, ie. any type of logging can be enabled/disabled. Note that this capability is not yet exposed to the CLI directly. A few questions need to be answered: 

- [ ] Do we persist logging flags in preferences?
- [ ] If yes, _how_ are the logging flags/preferences exposed? I think we need to have `t2 config` (and it should match [npm config](https://docs.npmjs.com/cli/config)

Demos:

- With and without: `--loglevel=debug`
  
- https://www.youtube.com/watch?v=2dj0ylLgAf8
- https://www.youtube.com/watch?v=nz6x_Gr8eKA
- https://www.youtube.com/watch?v=nz6x_Gr8eKA

Notes: 

- npmlog is presently using a branch on my fork, pending acceptance of https://github.com/npm/npmlog/pull/34
- There may be more places that the spinner can be added, so please feel free to suggest
- There are currently no `http` or `trace` logs, but now that this is in place, I will spend some time implementing those.

### CODE REVIEW

File: bin/tessel-2.js [hunk 1]
```diff
@@ -77,42 +79,50 @@ function makeCommand(commandName) {
...
   return controller[methodName](opts)
     .then(module.exports.closeSuccessfulCommand, module.exports.closeFailedCommand);
 }
 
-function callControllerCallback(methodName) {
-  return function(opts) {
-    return callControllerWith(methodName, opts);
-  };
-}
-
```
Comment (rwaldron) [inline bin/tessel-2.js]: No longer used. All calls consolidated to `callControllerWith`

File: bin/tessel-2.js [hunk 2]
```diff
@@ -77,42 +79,50 @@ function makeCommand(commandName) {
...
 parser.command('install-drivers')
-  .callback(function() {
+  .callback(options => {
+    if (options.loglevel !== log.level()) {
+      log.level(options.loglevel);
+    }
```
Comment (rwaldron) [inline bin/tessel-2.js]: This is implemented for all `callback(...)`. Whatever args handling we replace nom-nom with should include a "global" callback.

File: bin/tessel-2.js [hunk 3]
```diff
@@ -77,42 +79,50 @@ function makeCommand(commandName) {
...
       cr = CrashReporter.on();
-    } else if (opts.off) {
+    } else if (options.off) {
       // t2 crash-reporter --off
       cr = CrashReporter.off();
     }
 
     // t2 crash-reporter --test
-    if (opts.test) {
+    if (options.test) {
```
Comment (rwaldron) [inline bin/tessel-2.js]: I renamed `opts` => `options` everywhere.

**Comment by rwaldron** [comment]
cc @HipsterBrown

**Comment by HipsterBrown** [comment]
Will review this tonight.

File: lib/controller.js [hunk 4]
```diff
@@ -66,11 +66,11 @@ controller.setupLocal = function(opts) {
 
 
 Tessel.list = function(opts) {
-
   return new Promise(function(resolve, reject) {
-
+    log.spinner.stop();
```
Comment (HipsterBrown) [inline lib/controller.js]: Is this just making sure the spinner is defaulted to stop before starting it, in case it was not stopped by `callControllerWith()`?

File: lib/discover.js [hunk 5]
```diff
@@ -5,14 +5,17 @@ var util = require('util');
 var Emitter = events.EventEmitter;
 
 // Third Party Dependencies
-var debug = require('debug')('discovery');
+// ...
```
Comment (HipsterBrown) [inline lib/discover.js]: Keeping this for a reason?

File: lib/discover.js [hunk 6]
```diff
@@ -5,14 +5,17 @@ var util = require('util');
...
 // Internal
 var lan = require('./lan-connection');
-var logs = require('./logs');
+var log = require('./log');
 var Tessel = require('./tessel/tessel');
 var usb = require('./usb-connection');
 
+function debug(message) {
+  log.debug(`(discovery) ${message}`);
+}
```
Comment (HipsterBrown) [inline lib/discover.js]: Smart 👍

File: lib/init.js [hunk 7]
```diff
@@ -176,7 +182,8 @@ module.exports = function(opts) {
   // Set interactive boolean off of CLI flags
   options = opts;
 
-  console.log('Initializing Tessel 2 Project...');
+  log.info('Initializing Tessel 2 Project...');
+  log.spinner.start();
```
Comment (HipsterBrown) [inline lib/init.js]: If this is started here, does `log.spinner.start()` need to be in all the async functions like `loadNpm`? Probably doesn't hurt to call it in each async function, just wondering.

File: lib/discover.js (hunk 5 above)
Comment (HipsterBrown) [inline lib/discover.js]: Nevermind, I see that it's a pattern in the other files as well. 👍

File: lib/tessel/access-point.js [hunk 8]
```diff
@@ -126,7 +128,10 @@ Tessel.prototype.createAccessPoint = function(opts) {
 
     var turnAccessPointOn = () => this.simpleExec(commands.turnAccessPointOn());
 
-    var commitAndClosePromise = () => new Promise((resolve, reject) => commitAndClose(this, status, resolve, reject));
+    var commitAndClosePromise = () => {
+      log.spinner.stop();
```
Comment (HipsterBrown) [inline lib/tessel/access-point.js]: Could this be called in `commitAndClose` since that's another async action that restarts a few system processes like `wifi`, `dnsmasq`, and `dhcp`, which could take a few moments worthy of a spinner.

File: lib/tessel/access-point.js [hunk 9]
```diff
@@ -146,12 +151,16 @@ Tessel.prototype.createAccessPoint = function(opts) {
...
     .then(() => {
+      log.spinner.stop();
+
       // When an AP exists, change the status to
       // reflect an "update" vs. "create":
       status = 'Updated Access Point successfully. ';
       return setupAccessPoint();
     })
     .catch(() => {
+      log.spinner.stop();
```
Comment (HipsterBrown) [inline lib/tessel/access-point.js]: `setupAccessPoint` calls this after the other async actions of `setAccessPoint` and `setLanNetwork`. Could this be removed so the spinner continues until the access point is actually finished being setup?

File: lib/tessel/access-point.js [hunk 10]
```diff
@@ -146,12 +151,16 @@ Tessel.prototype.createAccessPoint = function(opts) {
 
   return this.simpleExec(commands.getAccessPointSSID())
     .then(() => {
+      log.spinner.stop();
```
Comment (HipsterBrown) [inline lib/tessel/access-point.js]: Similar note as this comment -> https://github.com/tessel/t2-cli/pull/768/files#r66899373

File: lib/tessel/deploy.js [hunk 11]
```diff
@@ -333,14 +333,17 @@ exportables.sendBundle = function(tessel, opts) {
...
 
             // Log write
-            logs.info('Writing project to %s on %s (%d kB)...', memtype, tessel.name, bundle.length / 1000);
+            log.info('Writing project to %s on %s (%d kB)...', memtype, tessel.name, bundle.length / 1000);
+            log.spinner.start();
 
             // Calling receive to know when the process closes
             tessel.receive(remoteProcess, (err) => {
               if (err) {
                 return reject(err);
```
Comment (HipsterBrown) [inline lib/tessel/deploy.js]: Will the spinner stop if this `Promise` is rejected?

File: lib/tessel/update.js [hunk 12]
```diff
@@ -54,18 +54,20 @@ Tessel.prototype.updateOpenWRT = function(opts, image) {
   return this.fixOldUpdateScripts()
     .then(() => {
       return new Promise((resolve, reject) => {
-        logs.info('Updating OpenWRT (1/2)');
+        log.info('Updating OpenWRT (1/2)');
+        log.spinner.start();
         // Write the new image to a file in the /tmp dir
         this.connection.exec(commands.openStdinToFile(updatePath), (err, remoteProc) => {
           if (err) {
             return reject(err);
```
Comment (HipsterBrown) [inline lib/tessel/update.js]: Will the spinner stop if this `Promise`  is rejected?

File: lib/tessel/update.js [hunk 13]
```diff
@@ -77,20 +79,21 @@ Tessel.prototype.updateOpenWRT = function(opts, image) {
...
-          logs.info('Configuration is saved during update.');
+          log.info('Configuration is saved during update.');
           sysupgradeCommand = commands.sysupgrade(updatePath);
         }
 
-        // The USBDaemon will cut out or the SSH command will close        
+        // The USBDaemon will cut out or the SSH command will close
         this.connection.exec(sysupgradeCommand, (err, remoteProc) => {
           if (err) {
             return reject(err);
```
Comment (HipsterBrown) [inline lib/tessel/update.js]: Same notes about [rejected Promises](https://github.com/tessel/t2-cli/pull/768/files#r66899802) and stopping the spinner.

File: lib/tessel/wifi.js [hunk 14]
```diff
@@ -6,14 +6,14 @@
...
-var logs = require('../logs');
+var log = require('../log');
 var Tessel = require('./tessel');
 
 
 Tessel.prototype.findAvailableNetworks = function() {
-  logs.info('Scanning for available networks...');
   return this.simpleExec(commands.scanWiFi())
     .then(function wifiScanResults(result) {
+      log.spinner.stop();
```
Comment (HipsterBrown) [inline lib/tessel/wifi.js]: When does this spinner start?

File: lib/tessel/wifi.js [hunk 15]
```diff
@@ -63,6 +63,7 @@ Tessel.prototype.getWifiInfo = function() {
 
       return this.simpleExec(commands.getIPAddress())
         .then(function(ipResults) {
+          log.spinner.stop();
```
Comment (HipsterBrown) [inline lib/tessel/wifi.js]: When does this spinner start?

**Comment by HipsterBrown** [comment]
> Do we persist logging flags in preferences?

I say yes and agree with following the `npm config` pattern. This leaves room to grow into other config flags if needed.

**Comment by rwaldron** [comment]
After reading the review comments, I decided to test another approach. As a result, I've pushed a completely revised version of this patch.

**Comment by HipsterBrown** [comment]
Reviewing and smoke testing now.

File: bin/tessel-2.js [hunk 16]
```diff
@@ -41,6 +41,45 @@ if (!isRoot()) {
...
+        module.exports.closeFailedCommand
+      );
+  }
+
+  return cr;
+};
+
+
+controller.installDrivers = function() {
+  require('./tessel-install-drivers');
```
Comment (HipsterBrown) [inline bin/tessel-2.js]: Is this being required again for a reason? It's also on line 16:

```js
var drivers = require('./tessel-install-drivers');
```

File: lib/usb-connection.js [hunk 17]
```diff
@@ -23,7 +29,7 @@ try {
   // var VENDOR_REQ_IN  = usb.LIBUSB_REQUEST_TYPE_VENDOR | usb.LIBUSB_RECIPIENT_DEVICE | usb.LIBUSB_ENDPOINT_IN;
```
Comment (HipsterBrown) [inline lib/usb-connection.js]: Do we still need this?

File: lib/usb-connection.js [hunk 18]
```diff
@@ -23,7 +29,7 @@ try {
...
 } catch (e) {
   haveusb = false;
-  logs.err('WARNING: No usb controller found on this system.');
+  log.error('WARNING: No usb controller found on this system.');
 }
```
Comment (HipsterBrown) [inline lib/usb-connection.js]: I ran into this error while testing this patch - `[Error: Error: Module version mismatch. Expected 47, got 46.]`, which came from the internals of the `usb` module and lefts the process hanging until I force quit it with `kill -9`. That error didn't come through until I added `log.error(e)` above line 31. I'm still not sure why it was left hanging open though because this didn't happen on `master`. Maybe there should be a `Promise.reject()`? 

The solution for that error was to reinstall my `node_modules` by the way.

Comment (rwaldron) [inline lib/usb-connection.js]: Seems like an entirely unrelated problem that normal use can't possibly warrant handling for :P


that is... i think you changed node versions, npm installed, then changed back.

File: lib/usb-connection.js (hunk 17 above)
Comment (rwaldron) [inline lib/usb-connection.js]: I don't know, but that should be addressed in a different patch

File: lib/tessel/update.js (hunk 13 above)
Comment (rwaldron) [inline lib/tessel/update.js]: I've updated this so that both `closeSuccessfulCommand` and `closeFailedCommand` can stop the spinner

File: lib/usb-connection.js (hunk 18 above)
Comment (HipsterBrown) [inline lib/usb-connection.js]: Yea, I did something screwy with `nvm` when creating new bash sessions. The comment was mostly recognizing some missing error feedback around this part of logging. Could we also have a line that logs the error coming from the `try...catch` around `require('usb')`? Or would that be part of another patch?

File: bin/tessel-2.js (hunk 16 above)
Comment (rwaldron) [inline bin/tessel-2.js]: Pfft. I never even noticed that. I just copied this from the old code.

Comment (rwaldron) [inline bin/tessel-2.js]: Fixed!

File: lib/usb-connection.js (hunk 18 above)
Comment (rwaldron) [inline lib/usb-connection.js]: That's a fail fast error, I don't think we need to do anything

**Comment by HipsterBrown** [comment]
Overall, I love the simplification of spinner state management. This is going to be a very useful patch. 

LGTM 👍

**Comment by rwaldron** [comment]
@HipsterBrown thank you so much for the several-days-long review, I appreciate your help enormously <3
